app/broadcast_areas/create-postcode-areas-db.py

```diff
@@ -0,0 +1,212 @@
+    """
+    This function expects a list of lists of coordinates defined in degrees
+    """
+    for index, polygon in enumerate(polygons):
+        shapely_polygon = Polygon(polygon)
+
+        # Some of our data has points which are incredibly close
+        # together. In some cases they are close enough to be duplicates
+        # at a given precision, which makes an invalid topology. In
+        # other cases they are close enough that, when converting from
+        # one coordinate system to another, they shift about enough to
+        # create self-intersection. The fix in both cases is to reduce
+        # the precision of the coordinates and then apply simplification
+        # with a tolerance of 0.
+        simplified_polygon = wkt.loads(
+            wkt.dumps(shapely_polygon, rounding_precision=Polygons.output_precision_in_decimal_places - 1)
+        ).simplify(0)
+
+        if simplified_polygon.is_valid:
+            print(f"{indent}Polygon {index + 1}/{len(polygons)} is valid")  # noqa: T201
+            yield simplified_polygon
+
+        else:
+            invalid_polygons.append(shapely_polygon)
+
+            # We’ve found polygons where all the points line up, so they
+            # don’t have an area. They wouldn’t contribute to a broadcast
+            # so we can ignore them.
+            if simplified_polygon.area == 0:
+                print(f"{indent}Polygon {index + 1}/{len(polygons)} has 0 area, skipping")  # noqa: T201
+                continue
+
+            print(f"{indent}Polygon {index + 1}/{len(polygons)} needs fixing...")  # noqa: T201
+
+            # Buffering with a size of 0 is a trick to make valid
+            # geometries from polygons that self intersect
+            buffered = shapely_polygon.buffer(0)
+
+            # If the buffering has caused our polygon to split into
+            # multiple polygons, we need to recursively check them
+            # instead
+            if isinstance(buffered, MultiPolygon):
+                try:
+                    for sub_polygon in clean_up_invalid_polygons(buffered, indent="        "):
+                        yield sub_polygon
+                    continue
+                except TypeError:
+                    for sub_polygon in clean_up_invalid_polygons(buffered.geoms, indent="        "):
+                        yield sub_polygon
+                    continue
+            # We only care about the exterior of the polygon, not an
+            # holes in it that may have been created by fixing self
+            # intersection
+            fixed_polygon = Polygon(buffered.exterior)
+
+            # Make sure the polygon is now valid, and that we haven’t
+            # drastically transformed the polygon by ‘fixing’ it
+            assert fixed_polygon.is_valid
+            # assert isclose(fixed_polygon.area, shapely_polygon.area, rel_tol=0.001)
+
+            print(f"{indent}Polygon {index + 1}/{len(polygons)} fixed!")  # noqa: T201
```

Comment:
Given the noqa happens a lot, it's probably worth turning this warning off for the entire file:

https://flake8.pycqa.org/en/latest/user/options.html#cmdoption-flake8-per-file-ignores

Reply:
The area creation scripts have been referenced in flake8 file exclusions.